extensions/blocks/social-previews/upgrade.js

```diff
@@ -43,10 +59,79 @@ const SocialPreviewsUpgrade = function SocialPreviewsUpgrade() {
 						) }
 					</li>
 				</ul>
-				<Button isPrimary>{ __( 'Upgrade', 'jetpack' ) }</Button>
+
+				<Button
+					href={ href } // Only for server-side rendering, since onClick doesn't work there.
+					isPrimary
+					label={ __( 'Purchase a business plan to access social previews', 'jetpack' ) }
+					onClick={ autosaveAndRedirect }
+					target="_top"
+				>
+					{ __( 'Upgrade', 'jetpack' ) }
+				</Button>
 			</div>
 		</div>
 	);
 };
 
-export default SocialPreviewsUpgrade;
+export default compose( [
+	withSelect( select => {
+		const plan = select( 'wordpress-com/plans' ).getPlan( BUSINESS_PLAN );
+		const postId = select( 'core/editor' ).getCurrentPostId();
+		const postType = select( 'core/editor' ).getCurrentPostType();
+
+		const planPathSlug = plan?.path_slug;
+		const domain = getSiteFragment();
+
+		// The editor for CPTs has an `edit/` route fragment prefixed.
+		const postTypeEditorRoutePrefix = [ 'page', 'post' ].includes( postType ) ? '' : 'edit';
+
+		// Post-checkout: redirect back here.
+		const redirect_to = isSimpleSite()
+			? addQueryArgs(
+					'/' + compact( [ postTypeEditorRoutePrefix, postType, domain, postId ] ).join( '/' ),
+					{
+						plan_upgraded: 1,
+					}
+			  )
+			: addQueryArgs(
+					window.location.protocol + `//${ domain.replace( '::', '/' ) }/wp-admin/post.php`,
+					{
+						action: 'edit',
+						post: postId,
+						plan_upgraded: 1,
+					}
+			  );
+
+		const href =
+			planPathSlug &&
+			addQueryArgs( `https://wordpress.com/checkout/${ domain }/${ planPathSlug }`, {
+				redirect_to,
+			} );
+
+		return {
+			href,
+			trackViewEvent: () =>
+				void analytics.tracks.recordEvent( 'jetpack_editor_block_upgrade_nudge_impression', {
+					plan: planPathSlug,
+					block: 'social-previews',
+				} ),
+			trackClickEvent: () =>
+				void analytics.tracks.recordEvent( 'jetpack_editor_block_upgrade_click', {
+					plan: planPathSlug,
+					block: 'social-previews',
+				} ),
+		};
+	} ),
+	withDispatch( ( dispatch, { href, trackClickEvent } ) => ( {
+		autosaveAndRedirect: async event => {
+			event.preventDefault(); // Don't follow the href before auto-saving.
+
+			trackClickEvent();
+			await dispatch( 'core/editor' ).autosave();
+
+			// Using window.top to escape from the editor iframe on WordPress.com.
+			window.top.location.href = href;
+		},
+	} ) ),
```

Comment:
I'm pretty sure @retrofox has something similar going on in Premium Blocks. I wonder whether we're reinventing the wheel between our two projects?

Reply:
yes, `Save&Redirect` has been already tackled on https://github.com/Automattic/jetpack/pull/16611. I'm going to try to make an abstraction in order to make it reusable. Part of the code is ready to use it, such as the getUpgradeUrl() helper function.

Reply:
I've created a `useUpgradeFlow` hook, [This is the commit](4a07e88db171615ce2c3dde4dcdce82832301b12), and then added other ones in order to refact this PR. Please feel free to discard them.